### PR TITLE
[TPU Offload][FIX] Release Staging buffer and remove slicing in bucked ops.

### PR DESCRIPTION
**Pull request by juncgu-google**
# Description

1. Fixed: release staging buffer only after its save is done. Previously, we mark a request as finished_request when its gathers are done (in order to let vllm scheduler release its kv pages); the request will be added into finished_sending and its staging buffer will be released.
2. Removed slicing (`jax.lax.dynamic_slice_in_dim`) sub block_ids array from the large block_to_save array for bucked operations ( the slicing triggers compilation).

# Tests

`pytest -s -v tests/offload/tpu_offload_connector_worker_test.py`

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/offload/tpu_offload_connector.py [hunk 1]
```diff
@@ -1899,26 +1903,29 @@ def _bucketed_update_kv_caches(
         if num_blocks == 0:
             return kv_caches
         if num_blocks in BLOCK_SIZE_BUCKETS:
-            return update_kv_caches(kv_caches, kv_cache_slices, dst_blocks,
+            dst_blocks_arr = jnp.array(dst_blocks)
```
Comment (saikat-royc) [inline tpu_inference/offload/tpu_offload_connector.py:1916]: this needs to be fixed as well?

File: tpu_inference/offload/tpu_offload_connector.py [hunk 2]
```diff
@@ -2583,7 +2590,7 @@ def start_load_kv(self, fwd_ctx: "ForwardContext") -> None:
                     self.runner.kv_caches = self._bucketed_update_kv_caches(
                         self.runner.kv_caches,
                         raw_chunked_kv_on_tpu,
-                        jnp.array(dst_blocks),
+                        dst_blocks,
```
Comment (saikat-royc) [inline tpu_inference/offload/tpu_offload_connector.py:2607]: also needs a fix in L2599

File: tpu_inference/offload/tpu_offload_connector.py (hunk 1 above)
Comment (juncgu-google) [reply tpu_inference/offload/tpu_offload_connector.py:1916]: no, update_kv_caches gets the blocks_ids in jax.array.

**Comment by saikat-royc** [comment]
please rebase on top of https://github.com/vllm-project/tpu-inference/pull/1801
